### PR TITLE
Relayer v2 architecture redesign and new packet worker MVP

### PR DESCRIPTION
**Pull request by soareschen**
## Description

The `cargo doc` documentation for this PR is available at https://ibc-rs.netlify.app/ibc_relayer_framework/.

[Rendered ADR](https://github.com/informalsystems/hermes/blob/soares/relayer-next/docs/architecture/adr-012-ibc-relayer-v2.md)

[Concurrency Architectures](https://github.com/informalsystems/hermes/wiki/Relayer-V2-Concurrency-Architecturess) ([Excalidraw Diagrams](https://app.excalidraw.com/l/4XqkU6POmGI/9oNGlMca2wF))

______

### PR author checklist:
- [ ] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [ ] Added tests: integration (for Hermes) or unit/mock tests (for modules).
- [ ] Linked to GitHub issue.
- [ ] Updated code comments and documentation (e.g., `docs/`).
- [ ] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: relayer-framework-guide/docs/context-generic-programming/13-concrete-composition.md [hunk 1]
```diff
@@ -0,0 +1,532 @@
...
+from the concrete types. If we know that our application only need the
+`SimpleTime` trait, then there are more options out there that we can
+try out and switch easily.
+
+It is also worth noting that it doesn't matter whether the concrete types
+`AppContext` and `BasicPerson` can have private fields or public fields.
+Since the components do not have access to the concrete types, all concrete
+fields are essentially private and can only be exposed via trait methods.
+
+Finally, we define an `instances` module to encapsulate the witness of
```
[the remark this answers is not in the document]
Comment (soareschen) [reply relayer-framework-guide/docs/context-generic-programming/13-concrete-composition.md]: I have instead used the term "proof" and add more explanation to it here: https://ibc-rs.netlify.app/context-generic-programming/04-greeter-component.html#greeter-instance

File: relayer-framework-guide/docs/context-generic-programming/05-explicit-binding.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+# }
+#
+# trait QueryPersonContext: ErrorContext {
+#   fn query_person(&self, person_id: &PersonId) ->  Result<Person, Self::Error>;
+# }
+#
+fn greet<Context, Error>(context: &Context, person_id: &PersonId)
+  -> Result<(), Error>
+where
+  Context: QueryPersonContext<Error=Error>,
```
[the remark this answers is not in the document]
Comment (soareschen) [reply relayer-framework-guide/docs/context-generic-programming/05-explicit-binding.md]: Yeah naming can be hard. I have tried changing the name to `HasError`. Not sure if it makes more sense?

File: relayer-framework-guide/docs/context-generic-programming/16-querier-consumer.md [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+the trait also requires the `Sized` supertrait, which is already implemented
+by most types other than `dyn Trait` types, so that we can use `Self` inside
+other generic parameters.
+
+In the body of `PersonQuerierContext`, we define a `PersonQuerier` associated
+type, which implements the trait `PersonQuerier<Self>`. This looks a little
+self-referential, as the context is providing a type that is referencing back
+to itself. But with the dependency injection mechanism of the traits system,
+this in fact works most of the time as long as there are no actual cyclic
+dependencies.
```
[the remark this answers is not in the document]
Comment (soareschen) [reply relayer-framework-guide/docs/context-generic-programming/16-querier-consumer.md]: I have tried to add more detailed explanation to `PersonQuerier<Self>` here: https://ibc-rs.netlify.app/context-generic-programming/08-querier-consumer.html#generic-querier-consumer

File: relayer-framework-guide/docs/context-generic-programming/16-querier-consumer.md [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+by most types other than `dyn Trait` types, so that we can use `Self` inside
+other generic parameters.
+
+In the body of `PersonQuerierContext`, we define a `PersonQuerier` associated
+type, which implements the trait `PersonQuerier<Self>`. This looks a little
+self-referential, as the context is providing a type that is referencing back
+to itself. But with the dependency injection mechanism of the traits system,
+this in fact works most of the time as long as there is no actual cyclic
+dependencies.
+
```
[the remark this answers is not in the document]
Comment (soareschen) [reply relayer-framework-guide/docs/context-generic-programming/16-querier-consumer.md]: I have split up the example code and move the relevant parts downward.

File: relayer-framework-guide/docs/context-generic-programming/05-daytime-greeter.md [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+errors into the main error type by requiring additional `From` constraints.
+
+For instance, `DaytimeGreeter` does not need to be aware of whether the inner
+greeter component would raise a database error. From the `impl` definition,
+we can be confident that `DaytimeGreeter` itself cannot raise any sub-error
+other than `ShopClosedError`.
+
+## Multiple Type Bindings
+
+When specifying the constraints for indirect depedencies, we have to keep using
```
Comment (AlianBenabdallah) [inline relayer-framework-guide/docs/context-generic-programming/05-daytime-greeter.md:214]:
```suggestion
When specifying the constraints for indirect dependencies, we have to keep using
```

File: relayer-framework-guide/docs/context-generic-programming/04-greeter-component.md [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+figures out that `SimpleGreeter` requires `AppContext` to implement
+`PersonQuerier`. It would then try to automatically _resolve_ the
+dependency by searching for an implementation of `PersonQuerier`
+for `AppContext`. Upon finding the implementation, Rust "binds" that
+implementation with `SimpleGreeter` and returns it as an existential
+type that implements `Greeter<AppContext>`. As a result,
+we can treat the type returned from `app_greeter` as an abstract type,
+and "forget" the fact that `AppContext` implements `PersonQuerier`.
+
+This pattern of making use of Rust's trait system for depedency injection
```
Comment (AlianBenabdallah) [inline relayer-framework-guide/docs/context-generic-programming/04-greeter-component.md:239]:
```suggestion
This pattern of making use of Rust's trait system for dependency injection
```

File: relayer-framework-guide/docs/context-generic-programming/05-daytime-greeter.md (hunk 5 above)
Comment (soareschen) [reply relayer-framework-guide/docs/context-generic-programming/05-daytime-greeter.md]: Fixed. Thanks!

File: relayer-framework-guide/docs/context-generic-programming/04-greeter-component.md (hunk 6 above)
Comment (soareschen) [reply relayer-framework-guide/docs/context-generic-programming/04-greeter-component.md]: Fixed. Thanks!

**Comment by seanchen1991** [comment]
Had a discussion with @adizere and @Farhad-Shabani regarding the need to validate Cosmos-Substrate relaying as part of the packet worker refactor. That work is tracked in [https://github.com/ComposableFi/hyperspace/issues/4](https://github.com/ComposableFi/hyperspace/issues/4).

**Comment by soareschen** [comment]
This PR has been moved to the `main` branch of https://github.com/informalsystems/hermes-sdk.

**Pull request closed without merging**